# Lesson 16: Hunting Exoplanets In Space - Deploying A Prediction Model

---

---

---

---

### Teacher-Student Activities

In the last class, we learnt how to create line plots and scatter plots to visualise data. 

In this class, we will deploy a **prediction model** to predict which stars in the test dataset have a planet and which do not. 

**How prediction model works?**

**1. Learning:**
- The prediction model, through the training dataset, will learn the properties of a star that has a planet and also the properties of a star which does not have a planet. 

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C16/whj-galaxy-more+planet-apt-c16-01.png" height=400/>


**2. Testing and Prediction:**  

- Once the model has learnt the required properties, it will look for these properties in the test dataset and according to the properties it sees, it will predict whether a star has a planet or not.

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C16/whj-galaxy-less+planet-apt-c16-01.png" height=400/>

---

In this class, we will learn how to deploy the **Random Forest Classifier** model (a machine learning algorithm). 

**What is Machine Learning?**

- Machine learning is a branch of artificial intelligence in which a machine learns through data the different features on its own without being programmed by a computer programmer. 

```
TEACHER
    Without machine learning, we need to tell the machine what to do. 
```

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C16/whj-comp-boy-apt-c16.gif" height=400 >



With machine learning, the machine learning will figure out what to do itself, using the inputs we feed it.
    
- First, the algorithmic models are **trained** to perform some specific tasks by learning patterns from the data it sees, rather than through computer programming by a human expert.

- After training, our model is ready for making predictions for any new input.

```
TEACHER
   In this image, you can see that the machine is learning from the data 
   provided. for example, it has learned how a cup looks like from the alphabet 
   chart. Now, whenever the machine sees a cup again, it can easily recognize 
   that it is a cup.

```
<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C16/whj-machine-tarining-apt-c16-01.png" />


<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C16/whj-comp-tea+cup-apt-c16-01.png" />



- For our model, the machine will learn to recognise the flux values of stars having a planet on its own.
-  When a new dataset containing only the flux values of a star is shown to the machine, it will tell whether that star has a planet or not.

There are many machine learning models or algorithms to do this kind of prediction. One of them is **Random Forest Classifier**. 

---
**Random Forest Classifier:**

- It is used to classify outcomes into classes (or labels) based on some features. 
- For e.g., an animal that makes a 'Meow! Meow!' sound is classified (or labelled) as a cat, an animal that makes a 'Woof! Woof!' sound is classified as a dog, an animal which makes a hissing sound is classified as a snake etc.
- It uses the training data to learn how the given features relate to a particular class. 
- **For example:** A classifier can classify a given set of population into following two classes:
  - Those suffering from Diabetes.
  - Those not suffering from Diabetes.

  on the basis of different features of the population such as age, gender, blood pressure etc. 

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C16/whj-pateint-database-apt-c16-01.png" height=400/>

In this lesson, we will use the Random Forest Classifier model to classify whether a star has a planet or not. 
- The stars which have at least one planet are labelled as `2` 
- The stars not having a planet are labelled as `1`. 


Let's run all the codes in the code cells that we have already covered in the previous classes and begin this class from **Activity 1: The `value_counts()` Function** section.  You too run the code cells until the first activity.



---

#### Loading The Datasets

Create a Pandas DataFrame every time you start a Jupyter notebook.

Dataset links (Don't click on them):

1. Train dataset 

   
2. Test dataset 
   
   

In [ ]:
# Loading both the training and test datasets.
import pandas as pd

exo_train_df = pd.read_csv('')
exo_test_df = pd.read_csv('')

In [ ]:
# Number of rows and columns in the DataFrames.
print(exo_train_df.shape)
exo_test_df.shape

(5087, 3198)


(570, 3198)

In the previous classes, we have already checked that the training dataset does not have any missing value. So, we can skip the missing values check part.

---

#### Activity 1: The `value_counts()` Function
Our prediction model should classify the stars either as `1` or `2`. Let's find out how many stars in the test dataset are classified as `1` and `2`.

To compute how many times a value occurs in a series, we use the `value_counts()` function.

In [ ]:
# Student Action: Count the number of times a value occurs in a Pandas series.
exo_test_df['LABEL'].value_counts()

1    565
2      5
Name: LABEL, dtype: int64

There are `565` stars which are classified as `1` and `5` stars classified as `2` which means only `5` stars have a planet. Interestingly, if our prediction model mindlessly classifies every star as `1`, then it is a very accurate model. Why?

Because the accuracy of a model is calculated as **a percentage of the correct predictions out of the total number of predictions**. In this case, the percentage of the correct predictions is

 $\frac{565\times100}{570} = 99.122$ %

Thus, without actually deploying a proper prediction model, we can predict the stars having a planet with 99% accuracy. 

This is **WRONG**! This is where we need to be careful. Because we have very imbalanced data. The ultimate goal of the Kepler space telescope is to detect exoplanets in outer space. Hence, a machine learning model, based on some data should also **correctly** detect stars having planets. This means a prediction model will be considered useful if it correctly detects almost all the stars having a planet.

So, the prediction model which always labels every star as `1` is useless. Because it must detect almost all the stars having a planet.

Now, we are going to deploy the Random Forest Classifier model so that it can detect all the five (or at least three) stars having a planet.

---

### Random Forest Classifier^ 

A Random Forest is a collection (a.k.a. ensemble) of many decision trees. A decision tree is a flow chart which separates data based on some condition. If a condition is true, you move on a path otherwise, you move on to another path.


For e.g., in case of finding a star having a planet, you can construct the following decision tree: 

<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-14/decision-tree.png' width=600>

You could ask a question whether there is decrease in the flux values of a star. If the answer is no, then it clearly means the star does not have a planet. However, if the answer is yes, then you could ask another question to check whether the decrease is periodic or not. Again, if the answer is no, then the star does not have a planet. Otherwise, it has a planet.

This is one of the examples of a decision tree. Based on a problem, the decision tree could get more and more complex. 

A collection of `N` number of trees is a random forest wherein each tree gives some predicted value (in this case either class `1` or class `2`). 


<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-14/rfc-image.jpg' width=800>

<h3><b>For example:</h3></b>

```
TEACHER
   For example, a random forest model is used to classify whether an item is a 
   pen or a pencil. There are 3 decision trees, each predicting a class for the 
   item. The final predicted value is the majority class, which in this case is 
   Class 2 i.e. Pen.
```

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C16/Untitled-13-01.png" height=400>




The final predicted value is the majority class, i.e, the class that is predicted by the most number of decision trees in a random forest.

For the time being, just consider the Random Forest Classifier as some kind of a black-box which classifies data into different classes (in this case, either class `1` or class `2`) by learning the properties of every class through a training dataset. 

---

#### Activity 2: Importing `RandomForestClassifier`

We need to import a module called `RandomForestClassifier` from a package called `sklearn.ensemble`. The `sklearn` (or **scikit-learn**) is a collection of many machine learning modules. Almost every machine learning algorithm can be directly applied without a knowledge of math using the **scikit-learn** library. It is kind of a plug-and-play device.

You can read about it in the link provided in the **Activities** section under the title **`scikit-learn` - Random Forest Classifier**


In [ ]:
# Teacher Action: Import the required modules from the 'sklearn' library.
# Import the 'RandomForestClassifier' module from the 'sklearn.ensemble' library.
from sklearn.ensemble import RandomForestClassifier

---

#### Activity 3: Target & Feature Variables Separation

The `RandomForestClassifier` module has a function called `fit()` which takes two inputs: 
1. The first input is the collection of feature variables.
2. The second input is the target variable.  

Let us try to understand what are Feature variables and Target Variables.

**Features and Target:**

<h3><font color=red><b>
Have you ever came across any facial recognition system? <img src="https://curriculum.whitehatjr.com/APT+Asset/Final+C2+images/question.jpg" height=25/></b></h3>

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C16/whj-face-recognition-apt-c16-01.png"/>


Face recognition is a biometric identification process which is used to authenticate a person using its **facial features**. The system examines various facial **features** such as eyes, nose and lips to authenticate a target user.  

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C16/eyenoselip.png" height=400/>

Similarly, the machine learning algorithm uses various features to predict the outcome. 

```
NOTE for the TEACHER:
   The teacher can also give following example if the student is still
   not able to understand the concept of features and target. 

  Example: An algorithm may analyse different features of an apple like size, 
  color, taste etc to uniquely identify an apple. 
  In this case, 
  features --> size, color, taste
  target   --> 'apple' or 'not apple'
```


For our dataset, we will use flux values of a star i.e.`FLUX.1` to `FLUX.3197` columns to find out whether that star has a planet or not.

 
**Feature Variables:** The features are those variables which describe the features or properties of an entity.

**Target variable:** The variable which needs to be predicted is called a target variable.



For our dataset, 
- `FLUX.1` to `FLUX.3197` are feature variables. Hence, the values stored in these columns are the features of a star in exoplanets dataset.
- `LABEL` is the target variable because the prediction model needs to predict which star belongs to which class in the test dataset. Hence, the values stored in the `LABEL` column are the target values. 

<img src="https://curriculum.whitehatjr.com/APT+Asset/APT+C16/WHJ-feature+variable-target-apt-c16-01.png"/>

So, we need to extract the target variable and the feature variables separately from the training dataset. 

Let's store the feature variables in the `x_train` variable and the target variable in the `y_train` variable. We will separate the features using the `iloc[]` function. 

Recall the syntax for `iloc[]` function.

**Syntax:** 

`dataframe_name.iloc[row_position_start : row_position_end, column_position_start : column_position_end]`


We need all the rows from the training set. So, inside the `iloc[]` function, we will enter the colon (`:`) sign to get all the rows. We do not need the first column, i.e., the `LABEL` column. Therefore, inside the `iloc[]` function, as part of column indexing, enter `1` as the starting index followed by the colon (`:`) sign to include the rest of the columns from the training dataset.


In [ ]:
# Student Action: Extract the feature variables from the training dataset using the 'iloc[]' function.
x_train = exo_train_df.iloc[:, 1:]
x_train.head()

FLUX.1   FLUX.2   FLUX.3  ...  FLUX.3195  FLUX.3196  FLUX.3197
0    93.85    83.81    20.10  ...      61.42       5.08     -39.54
1   -38.88   -33.83   -58.54  ...       6.46      16.00      19.93
2   532.64   535.92   513.73  ...     -28.91     -70.02     -96.67
3   326.52   347.39   302.35  ...     -17.31     -17.35      13.98
4 -1107.21 -1112.59 -1118.95  ...    -384.65    -411.79    -510.54

[5 rows x 3197 columns]

Now, let's get only the target variables from the training dataset.

In [ ]:
# Student Action: Using the 'iloc[]' function, retrieve only the first column, i.e., the 'LABEL' column from the training dataset.
y_train = exo_train_df.iloc[:, 0] 
y_train.head()

0    2
1    2
2    2
3    2
4    2
Name: LABEL, dtype: int64

---

#### Activity 4: Fitting The Model

Now that we have separated the feature and target variables for deploying the `RandomForestClassifier` model, let's train the model with the feature variables using the `fit()` function. The steps to be followed are described below.

1. First, call the `RandomForestClassifier` module with inputs as `n_jobs=-1` and `n_estimators=50`. Store the function in a variable with the name `rf_clf`.

  For the time being, ignore the reason behind providing the `n_jobs=-1` parameter as an input.

  ```python
  rf_clf = RandomForestClassifier(n_jobs=-1, n_estimators=50)
  ```

  The `n_estimators` parameter defines the number of decision trees in a Random Forest. Therefore, `n_estimators=50` means that the forest contains `50` decision trees. If `n_estimators` parameter is not defined by a user, then by default, the forest contains `100` decision trees.

2. Call the `fit()` function with `x_train` and `y_train` as inputs.

  ```python
  rf_clf.fit(x_train, y_train)
  ```
3. Call the `score()` function with `x_train` and `y_train` as inputs to check the accuracy score of the model. This step is actually not required. If you wish, you can skip this step.
  
  ```
  rf_clf.score(x_train, y_train)
  ```


In [ ]:
# Teacher Action: Train the 'RandomForestClassifier' model using the 'fit()' function.

# 1. First, call the 'RandomForestClassifier' module with inputs as 'n_jobs = - 1' & 'n_estimators=50'. Store it in a variable with the name 'rf_clf'.
# For the time being, ignore the reason behind providing the 'n_jobs=-1' parameter as an input.
rf_clf = RandomForestClassifier(n_jobs=-1, n_estimators=50)

# 2. Call the 'fit()' function with 'x_train' and 'y_train' as inputs.
rf_clf.fit(x_train, y_train)

# 3. Call the 'score()' function with 'x_train' and 'y_train' as inputs to check the accuracy score of the model.
rf_clf.score(x_train, y_train)

0.9998034204835856

As you can see, we have deployed the `RandomForestClassifier` model with an accuracy of 100% (the number `1.0` signifies 100% accuracy).


---

#### Activity 5: Target & Feature Variables From Test Dataset^^

Now we need to make predictions on the test dataset. So, we just need to extract feature variables from the test dataset `exo_test_df` using the `iloc[]` function.

In [ ]:
# Student Action: Using the 'iloc[]' function, extract the feature variables from the test dataset.
x_test = exo_test_df.iloc[:, 1:]
x_test.head()

FLUX.1   FLUX.2   FLUX.3  ...  FLUX.3195  FLUX.3196  FLUX.3197
0   119.88   100.21    86.46  ...      35.78     269.43      57.72
1  5736.59  5699.98  5717.16  ...   -2366.19   -2294.86   -2034.72
2   844.48   817.49   770.07  ...    -162.68     -36.79      30.63
3  -826.00  -827.31  -846.12  ...    -120.81    -257.56    -215.41
4   -39.57   -15.88    -9.16  ...     -61.98     -69.34     -17.84

[5 rows x 3197 columns]

Let's also extract the target variable from the test dataset `exo_test_df` so that we can compare the actual target values with the predicted values later.

In [ ]:
# Student Action: Using the 'iloc[]' function, extract the target variable from the test dataset.
y_test = exo_test_df.iloc[:, 0]
y_test.head()

0    2
1    2
2    2
3    2
4    2
Name: LABEL, dtype: int64

---

#### Activity 6: The `predict()` Function^^^

Now, let's make predictions on the test dataset by calling the `predict()` function with the features variables of the test dataset as an input.

**Syntax of predict() function:**

```python
model.predict(data)
```
 where, `data` is a set of feature variables.

In [ ]:
# Student Action: Make predictions on the test dataset by using the 'predict()' function.
y_predicted = rf_clf.predict(x_test)
y_predicted

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

The `predict()` function returns a NumPy array of the predicted values. You can verify it using the `type()` function.

In [ ]:
# Student Action: Using the 'type()' function, verify that the predicted values are obtained in the form of a NumPy array.
type(y_predicted)

numpy.ndarray

The actual target values are stored in a Pandas series. So, for the sake of consistency, let's convert the NumPy array of the predicted values into a Pandas series.

In [ ]:
# Student Action: Convert the NumPy array of predicted values into a Pandas series.
y_predicted = pd.Series(y_predicted)
y_predicted.head()

0    1
1    1
2    1
3    1
4    1
dtype: int64

Now, let's count the number of stars classified as `1` and `2`.

In [ ]:
# Student Action: Using the 'value_counts()' function, count the number of times 1 and 2 occur in the predicted values.
y_predicted.value_counts()

1    570
dtype: int64

As you can see, we did not get the expected results. The model should have classified all the stars having a planet as `2`. Ideally, the Random Forest Classifier model should have classified `565` values as `1` and the remaining `5` values as `2`. 

In this case, even though the accuracy of a prediction model is high but according to the problem statement, it is not giving the desired result. Hence, **accuracy alone is not the metric to test the efficacy of a prediction model.** 

In the next class, we will try to investigate why Random Forest Classifier failed to classify even a single star as `2`. Based on the investigation, we will try to improve the model and then deploy it again.

---

### Additional Activities

The activities starting from this point are optional. Please do these activities **ONLY** if you have time to spare in the class. Otherwise, skip to the **Wrap-Up** section. The additional activities will not be available in the class copy of the notebook. You will have to manually add these activities in the class copy by adding new text and code cells.

Moreover, you don't have to do all the additional activities. Depending on the availability of time in a class, you can choose the number of additional activities to perform from this collection. 

---

#### Activity 1: Multiples

Given $n$ and $m$, write a function to print first $m$ multiples of a number $n$ without using any loop. E.g.,

```
Input: n = 2, m = 3
Output: 2 4 6 

Input: n = 3, m = 4
Output: 3 6 9 12
```

In [ ]:
# Solution
import numpy as np
def m_multiples_of_n(n, m):
  multiples = np.arange(n, (n * m) + 1, n)
  print(multiples)

m_multiples_of_n(2, 3)
m_multiples_of_n(3, 4)

[2 4 6]
[ 3  6  9 12]


---

#### Activity 2: Centered Penatognal Number

Write a function that takes a positive integer and calculates how many dots exist in a pentagonal shape around the centre dot on the $N^{\text{th}}$ iteration.

In the image below you can see the first iteration is only a single dot. On the second, there are 6 dots. On the third, there are 16 dots, and on the fourth, there are 31 dots.

<img src='https://drive.google.com/uc?id=1wn1u68GW15KGIz_MWsioUp74B9BPghuP'> 

So for first 21 positive integers, there should exist 1, 6, 16, 31, 51, 76, 106, 141, 181, 226, 276, 331, 391, 456, 526, 601, 681, 766, 856, 951, 1051 dots.

In [ ]:
# Solution
def centered_pentagonal_dots(num_pentagons):
  num = 1
  difference = 5
  num_pentagons_list = []
  while num_pentagons > 0:  
    num_pentagons_list.append(num)
    num = num + difference
    difference += 5
    num_pentagons -= 1
  # return num_pentagons_list
  return num_pentagons_list[-1]

print(centered_pentagonal_dots(4))
print(centered_pentagonal_dots(10))
print(centered_pentagonal_dots(21))

31
226
1051


---

---

---